### ISSUE TITLE
Encoding of data when publishing

### ISSUE THREAD

**Issue by blowmage** [GoogleCloudPlatform/google-cloud-ruby#1536]
In #1530 there are several discussions of how to deal with publishing data that is not already binary encoded. The current behavior is to match the REST implementation's use of Base64 and to use `#force_encoding`. @lukesneeringer has asked for the following:

1) Change the implementation to not allow data that is not already binary encoded. That would mean to raise an error when publishing data that was not already using the encoding `ASCII-8BIT`.
2) If the ruby library continues to be liberal in the data is accepts to switch to use `encode!` which will raise when characters are not able to be represented. So instead of the UTF-8 string `"👍"` being represented as the bytes `"\xF0\x9F\x91\x8D"`, an `Encoding::UndefinedConversionError` error would be raised.

**Comment by geigerj** [comment]
In the short term at least, I would prefer option 1 for consistency with the Python/Node client libraries. Consistency is especially important here because a Publisher may be using a different client library than the subscriber, and we don't want to imply by auto-encoding the publish request that the Subscriber will know how to decode it. (We can also relax this restriction after the first iteration of the client if it winds up being something that trips users up.)

**Comment by thagomizer** [comment]
If we have to pick between the two I would prefer #2. AFAIK we haven't had any issues with the current handling of encoding and raising an error will alert users if there may be an issue.

**Comment by blowmage** [comment]
I should have added a third option to keep things as they are. That is my preference, FWIW. :)

All strings or IO objects can be expressed as a series of bytes, which is what we are currently sending as the message data. All `#force_encoding` is doing now is taking the existing list of bytes and changing the encoding associated with it so it can be sent to the RPC. It isn't modifying the bytes.

**Comment by blowmage** [comment]
Here is another way to look at the existing implementation. Say the user publishes the UTF-8 string "👍". That string has the following bytes:

```ruby
"👍".bytes #=> [240, 159, 145, 141]
```

If you pack each byte number as an 8-bit unsigned byte, you get the following:

```ruby
[240, 159, 145, 141].bytes.pack("C*") #=> "\xF0\x9F\x91\x8D"
```

This is the same as calling `#force_encoding`:

```ruby
"👍".force_encoding "BINARY" #=> "\xF0\x9F\x91\x8D"
```

So the current behavior is to take whatever bytes the user provides, regardless of encoding, and publish them. I think that is a viable option to be considered.

**Comment by lukesneeringer** [comment]
@blowmage, thank you for separating out this issue for discussion. I do want to /cc @tbkneeland because this has been an issue that the Pub/Sub team has been passionate about.

I also do want to copy over my original commentary in full; I will follow up momentarily with responses.

------------

Summoning @remi and @tbkneeland; we need discussion here. :-)

So, disclaimer: I am a Python guy so I have Python biases. That said, I think this should be an error rather than an implicit conversion. Same for line 128 below.
This is a hard spot, though, because I know that being implicit and Postel's Law are important in Ruby.

Here is why I think this might be important enough to break Ruby convention: Bear in mind that with Pub/Sub, you do not know anything about the client on the other side. They may not (in fact, likely will not) be using this library, or Ruby at all. You can not assume _anything_ about the folks on the consuming side of this. Well, that is not _quite_ true: You can basically assume that the client will base64 decode the message, and nothing else.

I am doing a quick bit of research on string encoding in Ruby (since I do not know much about the ins and outs of this topic in Ruby; note that I _am_ fluent on the topic more generally -- an entire chapter of my book was devoted to this issue), and as best as I can tell, Ruby [basically has the same problems that everyone else has][1]. Basically, strings have an explicit `encoding`, but it might be wrong, especially if you read in from a file.

This merits discussion. But I _think_ that users need to send the right thing, and we should actually complain loudly rather than coerce for them, because the fundamental awareness here is so critical.

  [1]: http://www.justinweiss.com/articles/3-steps-to-fix-encoding-problems-in-ruby/

Also, if we are going to encode strings for users (needs discussion as stated above), we _*definitely*_ should not be using `force_encoding`. We should be using `encode!`.

According to the Ruby documentation, [encode][3] actually transcodes the string to the destination encoding, while [force_encoding][2] simply changes what the string says its encoding is. In other words, if the former is like translating from English to Greek, the latter is like _claiming that_ the text is Greek.

The fact that our own (excellent) developers got this wrong pushes me even further toward wanting to avoid implicit conversion of encodings, though. Our users _will_ send data and wonder why they are getting something else on the other end.

  [3]: https://ruby-doc.org/core-2.4.0/String.html#method-i-encode-21
  [2]: https://ruby-doc.org/core-2.4.0/String.html#method-i-force_encoding

**Comment by lukesneeringer** [comment]
Apparently my definition of "momentarily" was different than that of most dictionaries. Flying, what can you do?

**Comment by lukesneeringer** [comment]
Okay, now sitting down to write a real reply to this. :-)

First, a quick one: I said (based on naive looks at Stack Overflow) to use `.encode('BINARY')`, but I see you are saying `.encode('ASCII-8BIT')`. You are the expert on how Ruby encodings are spelled, so I completely defer to you. :-)  If you see references to `BINARY` in anything I write (past or future) you may consider them to be interchangeable in terms of what I say vs. what you do.

Next let's talk `#force_encode` versus `#encode!`. Disclaimer: I apologize if my comments came across as a slight. I did not intend them as such and I hold you in high regard. :-) I am pretty sure, however, that `#force_encode` is not correct. I could see an argument that it matters a little less if you are encoding to an opaque bytestring, but I am pretty sure there are probably cases where it does matter. One example: if the encoding that a text string actually is turns out to be different from its actual encoding, that _definitely_ needs to be an error, and `#encode!` seems to be to be the correct way to catch that.

> we chose `#force_encoding` to try to preserve the data that the user presents.

I am puzzled by this argument. `#force_encode`, unless I fundamentally misunderstand it, essentially replaces the string's encoding declaration with `ASCII-8BIT` even if it is not an 8-bit ASCII string. That seems like a loss of data to me.

My overarching concern here is: if I start with a string `foo` (with any encoding) in Ruby, and it goes to Pub/Sub, then on any language on the other end, I should be able to call `.decode` (or whatever the appropriate equivalent is) with the same encoding and end up with the same string.

Now, it is possible that there is a separate case where `#encode!` also does the wrong thing, but in a more subtle way: if the actual encoding does not match the declared encoding, but the encode _works_, for example. That brings us to the next question...

Should we continue to be liberal about what we accept, and try to do the right thing? I am thinking the answer is probably "no". I agree that it is annoying to have to convert the string yourself, but I also kind of feel like forcing the user to do that probably heads off other subtle bugs, or at least makes them easier to track down. The Pub/Sub team's contention is that it matters that _users explicitly understand_ that Pub/Sub sends bytes, and not text. The more I think through this problem, the more I think that position is correct.

On the other hand, if we only accept binary blobs, then we do not have to figure out which encoding method to use! ;-)

P. S. If we do go with the "raise an exception if we do not like your encoding" method, I expect we should probably accept _both_ `BINARY` and `ASCII-8BIT`. That might be wrong, though, if `BINARY` is not what I think it is.

**Comment by blowmage** [comment]
> First, a quick one: I said (based on naive looks at Stack Overflow) to use .encode('BINARY'), but I see you are saying .encode('ASCII-8BIT').

When encoding, the `BINARY` is an alias for `ASCII-8BIT` and can be used interchangeably.

```ruby
# using BINARY
source = "👍".force_encoding "BINARY"
source.encoding #=> #<Encoding:ASCII-8BIT>

# using ASCII-8BIT does the same thing
source.force_encoding "ASCII-8BIT"
source.encoding #=> #<Encoding:ASCII-8BIT>

# get the encoding from Ruby constants
Encoding::BINARY #=> #<Encoding:ASCII-8BIT>
Encoding::ASCII_8BIT #=> #<Encoding:ASCII-8BIT>
```

**Comment by blowmage** [comment]
> On the other hand, if we only accept binary blobs, then we do not have to figure out which encoding method to use! ;-)

The more I think about this, the more I'm convinced the current implementation is probably correct. The thing is, any passed in String or IO object is _already_ a binary blob. But it is a binary blob with an encoding. What we are doing is taking the bytes from that blob and passing them to Pub/Sub. For now, let's leave the issue of `#force_encoding` and `#encode!` aside and focus on the behavior. Let's assume some data encoded as UTF-8 bytes, and some different data encoded as UTF-16 bytes. We should take the bytes as-is and send them through. Here is how that would look:

```ruby
utf8_source = "hi"
utf8_source.bytes #=> [104, 105]
utf8_bytes = utf8_source.bytes.pack("C*") #=> "hi"

utf8_source.encoding #=> #<Encoding:UTF-8>
utf8_bytes.encoding #=> #<Encoding:ASCII-8BIT>

utf16_source = "hi".encode! "UTF-16"
utf16_source.bytes #=> [254, 255, 0, 104, 0, 105]
utf16_bytes = utf16_source.bytes.pack("C*") #=> "\xFE\xFF\x00h\x00i"

utf16_source.encoding #=> #<Encoding:UTF-16 (dummy)>
utf16_bytes.encoding #=> #<Encoding:ASCII-8BIT>
```

We don't care about the source's encoding, we only care about the bytes. The way the bytes are represented are as a separate binary encoded string. Is this something we can agree on? If so, then I will return to the discussion of the way to get there in my next comment.

**Comment by blowmage** [comment]
If we can agree that we should accept the bytes that are provided, then we can return to the discussion of the best way to accomplish this. `#force_encoding` leaves the source bytes as is, which is what we want, but mutates the source's encoding, which is likely something we don't want. `#encode!` mutates both the source bytes and encoding, converting it from the encoding it was to the new encoding, which is something I believe we don't want.

One issue with the existing implementation is that we are calling `#force_encoding` directly on the source, which while does not mutate the source bytes, does mutate the source encoding:

```ruby
source = "👍"
source.encoding #=> #<Encoding:UTF-8>
source.force_encoding "BINARY"
source.encoding #=> #<Encoding:ASCII-8BIT>
```

One way around this is to duplicate the source before changing the encoding of the bytes:

```ruby
source = "👍"
bytes = source.dup.force_encoding "BINARY"

source.encoding #=> #<Encoding:UTF-8>
bytes.encoding #=> #<Encoding:ASCII-8BIT>
```

The handling of the bytes is why I said the current implementation is probably correct. But we are not duplicating the source before extracting the bytes, so we are mutating the source's encoding. But I think if we add that call to `.dup` we can fix that. Thoughts?

**Comment by blowmage** [comment]
This comment is a side note about how we signal to the user that we are returning bytes. Both BigQuery and Spanner return bytes from their respective data stores, and the library wraps the binary encoded string in a StringIO object to differentiate it from other strings. I think that is a change we should make in Pub/Sub for similar reasons. So instead of `Message#data` returning a binary encoded string, it would return a StringIO object of the binary encoded string. That, along with updated documentation, should hopefully signal to the user that how the message data is being handled.

**Comment by blowmage** [comment]
My apologies for confusing the threading of the discussion a bit. I wanted to make this point before the side not about StringIO:

If we still really don't like the use of `#force_encoding` in `bytes = source.dup.force_encoding "BINARY"`, we could remove the method call completely and replace it with `bytes = source.bytes.pack("C*")`. Both are doing the exact same thing.

The reason I advocate for the use of `.dup.force_encoding "BINARY"` over `.bytes.pack("C*")` is that I suspect the former is a more efficient in its implementation. I can create some performance benchmarks on both approaches if that is wanted.

**Comment by lukesneeringer** [comment]
> If we can agree that we should accept the bytes that are provided, then we can return to the discussion of the best way to accomplish this. `#force_encoding` leaves the source bytes as is, which is what we want, but mutates the source's encoding, which is likely something we don't want. `#encode!` mutates both the source bytes and encoding, converting it from the encoding it was to the new encoding, which is something I believe we don't want.

This is a good thread.

So, I intuitively understand how `#encode!` mutates the bytes and encoding if you, say, take a latin1 string and encode it to utf-8. What happens exactly when going to binary?

I know you would _definitely_ end up with an error if you ended up with bytes that were invalid for the encoding. (I assert we want this.) If it is valid, what exactly happens?

> This comment is a side note about how we signal to the user that we are returning bytes. Both BigQuery and Spanner return bytes from their respective data stores, and the library wraps the binary encoded string in a StringIO object to differentiate it from other strings. I think that is a change we should make in Pub/Sub for similar reasons.

Yes, definitely.

> The reason I advocate for the use of `.dup.force_encoding "BINARY"` over `.bytes.pack("C*")` is that I suspect the former is a more efficient in its implementation. I can create some performance benchmarks on both approaches if that is wanted.

No need to convince me on this part. I defer to your expertise.

**Comment by blowmage** [comment]
> So, I intuitively understand how `#encode!` mutates the bytes and encoding if you, say, take a latin1 string and encode it to utf-8. What happens exactly when going to binary?

`#encode!` will raise an error for any encoded character that does not fit into ASCII. So assuming latin1 is compatible with 8-bit ASCII, no error will be raised. (I don't know for sure they are compatible, I just assume it.)

Here is an example of a UTF-16 string before and after having `.encode! "BINARY"` called on it:

```ruby
source = "hello".encode! "UTF-16"
source.bytes #=> [254, 255, 0, 104, 0, 101, 0, 108, 0, 108, 0, 111]
bytes = source.bytes.pack "C*" #=> "\xFE\xFF\x00h\x00e\x00l\x00l\x00o"

ascii_source = source.encode! "BINARY"
ascii_source.bytes #=> [104, 101, 108, 108, 111]
ascii_bytes = ascii_source.bytes.pack "C*" #=> "hello"
```

As you can see, calling `.encode! "BINARY"` changes the bytes in the blob. But since all the characters in the UTF-16 string can be represented in ASCII no errors were raised.

**Comment by lukesneeringer** [comment]
Okay, this is useful.

To put it in perspective, the requirement in Python is to take `str` objects and encode them into `bytes` (e.g. `s.encode('utf8')`); similarly, in Node, we only take `Buffer` objects. My understanding is that Java and Go are similar.

So...that is the equivalent of _what_ exactly here?

**Comment by blowmage** [comment]
> So...that is the equivalent of what exactly here?

Its hard for me to say because each environment is so different. In Java you have a Buffer, but you also have a ByteBuffer, which would probably be more useful for Pub/Sub. In Ruby we have a StringBuffer, not a separate ByteBuffer or CharBuffer.

A byte in Ruby is an instance of a String object. A character in Ruby is also an instance of a String object. Ruby doesn't have separate classes for these constructs. A byte is a single character string with he encoding ASCII-8BIT. A character is a single character string with an encoding. The character object may be 4 bytes long, but it is still considered a single character. So a StringBuffer could be used to collect characters, or bytes, or multi-character strings.

```ruby
byte = [42].pack "C" #=> "*"
byte.bytes #=> [42]
byte.chars #=> ["*"]
byte.encoding #=> #<Encoding:ASCII-8BIT>
byte.class #=> String

char = "👍" #=> "👍"
char.bytes #=> [240, 159, 145, 141]
char.chars #=> ["👍"]
char.encoding #=> #<Encoding:UTF-8>
char.class #=> String
```

The way binary data is dealt with in ruby is as a String object with the encoding ASCII-8BIT. Its a special case String. So the way I read the requirements for Pub/Sub message data is to take the data, format it as binary data in such a way that we do not modify the data given, and sent it to the API. In other words, we accept the binary blob we are given, and strip away the encoding.

**Comment by tbkneeland** [comment]
As Luke said: Yes, the CP/S team is very passionate about these encoding / interop issues... but I'll defer commenting on this until after reviewing the Publisher/Subscriber PRs

**Comment by blowmage** [comment]
I believe this is resolved by #1564.
